### DISCUSSION TITLE
Keep-alive connection lifetime

### DISCUSSION THREAD

**Discussion by lazitski-aliaksei** [encode/httpx#2077] [Q&A] [answered]
Is there are way to set max lifetime of the keep-alive connection?

For example, such setting may be helpful for clien-side loadbalancing in k8s, i.e. if connections got closed and opened again after some time they will end up in connecting to different instances of requested service.

**Comment by lazitski-aliaksei** [comment]
Example of such setting in akka-http https://doc.akka.io/docs/akka-http/current/common/timeouts.html#connection-lifetime-timeout

**Comment by lovelydinosaur** [comment, marked as answer]
I've escalated it into an issue, since it looks like it's not covered in the docs right now.
See issue https://github.com/encode/httpx/issues/2079 to track this.

You can control the keep alive expiry like this...

```python
limits = httpx.Limits(keepalive_expiry=30.0)
client = httpx.Client(limits=limits)
```

Thanks for the feedback - really great prompt to help us document this.

**Reply by lazitski-aliaksei** [reply]
Having keepalive_expiry documented will be great, but it looks like this setting controls a bit different thing from one I'm looking for: it controls how long keep-alive connection can be in idle state, while I'm looking for a way to close keep-alive connection even if it's actively used but only after some specified period of time, i.e. I want connection to be closed in 10 seconds after it was established no matter what.

**Reply by lovelydinosaur** [reply]
Is this the sort of thing you mean?...

```python
# Connections older than 60 seconds should *always* close as soon as they're idle.
limits = httpx.Limits(max_connection_lifespan=60.0)  # This setting doesn't actually exist
client = httpx.Client(limits=limits)
```

We don't have any support for that, and I'm *not sure?* if we'd want to consider it or not at this point in time.
But I could help guide you through how you'd implement it.

Hints...

* https://github.com/encode/httpcore/blob/b400d72cf15271c1f4abafa05dd8d82058493671/httpcore/_async/http11.py#L228-L240
* https://github.com/encode/httpcore/blob/b400d72cf15271c1f4abafa05dd8d82058493671/httpcore/_async/http11.py#L40-L45

**Reply by lazitski-aliaksei** [reply]
Yes, this is exactly the setting I'm looking for. I'll probably try to inject this behavior from my code to httpx. Looks like it's rarely used feature.

**Reply by lovelydinosaur** [reply]
Great. If you make progress with it then please do let us know how you get on.
